test/Microsoft.AspNet.Mvc.FunctionalTests/CorsTests.cs

```diff
@@ -0,0 +1,233 @@
+        [InlineData("HEAD")]
+        [InlineData("POST")]
+        public async Task ResourceWithSimpleRequestPolicy_Allows_SimpleRequests(string method)
+        {
+            // Arrange
+            var server = TestServer.Create(_provider, _app);
+            var client = server.CreateClient();
+            var origin = "http://example.com";
+
+            var requestBuilder = server
+                .CreateRequest("http://localhost/Cors/GetBlogComments")
+                .AddHeader(CorsConstants.Origin, origin);
+
+            // Act
+            var response = await requestBuilder.SendAsync(method);
+
+            // Assert
+            Assert.Equal(HttpStatusCode.OK, response.StatusCode);
+            var content = await response.Content.ReadAsStringAsync();
+            Assert.Equal("[\"comment1\",\"comment2\",\"comment3\"]", content);
+            var responseHeaders = response.Headers;
+            var header = Assert.Single(response.Headers);
+            Assert.Equal(CorsConstants.AccessControlAllowOrigin, header.Key);
+            Assert.Equal(new[] { "*" }, header.Value.ToArray());
+        }
+
+        [Theory]
+        [InlineData("GET")]
+        [InlineData("HEAD")]
+        [InlineData("POST")]
+        [InlineData("PUT")]
+        public async Task PolicyFailed_Disallows_PreFlightRequest(string method)
+        {
+            // Arrange
+            var server = TestServer.Create(_provider, _app);
+            var client = server.CreateClient();
+
+            // Adding a custom header makes it a non simple request.
+            var requestBuilder = server
+                .CreateRequest("http://localhost/Cors/GetBlogComments")
+                .AddHeader(CorsConstants.Origin, "http://example.com")
+                .AddHeader(CorsConstants.AccessControlRequestMethod, method)
+                .AddHeader(CorsConstants.AccessControlRequestHeaders, "Custom");
+
+            // Act
+            var response = await requestBuilder.SendAsync(CorsConstants.PreflightHttpMethod);
+
+            // Assert
+            // MVC applied the policy and since that did not pass, there were no access control headers.
+            Assert.Equal(HttpStatusCode.OK, response.StatusCode);
+            Assert.Empty(response.Headers);
+
+            // It should short circuit and hence no result.
+            var content = await response.Content.ReadAsStringAsync();
+            Assert.Equal(string.Empty, content);
+        }
+
+        [Fact]
+        public async Task SuccessfulCorsRequest_AllowsCredentials_IfThePolicyAllowsCredentials()
+        {
+            // Arrange
+            var server = TestServer.Create(_provider, _app);
+            var client = server.CreateClient();
+
+            // Adding a custom header makes it a non simple request.
+            var requestBuilder = server
+                .CreateRequest("http://localhost/Cors/EditUserComment?userComment=abcd")
+                .AddHeader(CorsConstants.Origin, "http://foo.com")
```

Comment:
Try to avoid any possibly real-looking hostname. `example.com` is always permitted (it is a specifically reserved domain name), as well as all its sub-domains.